### ISSUE TITLE
Using Streams as RedisValues

### ISSUE THREAD

**Issue by shravan2x** [StackExchange/StackExchange.Redis#866]
Note that this issue has nothing to do with the new Redis Streams feature.

I have a program that regularly publishes large JSON values over pub/sub channels. To do this, I use `JsonConvert.SerializeObject(...)`, which first serializes the object to a large `string` and then implicitly casts this to a `RedisValue` for transmission.

Is it possible to avoid this large string allocation each time by passing a stream instead? Basically, this would mean that a new implicit conversion be created between Stream and RedisValue. When the RedisValue is used, data from the stream is read.

**Comment by mgravell** [comment]
In principle, yes that's possible if the stream supports .Length and is fully rewound... But: I might have some better options. Let me check something...

**Comment by mgravell** [comment]
Note: mid-term, I think we could do something awesome here with "pipelines".

Right; to put some thoughts together: my guess here is that you're basically going to spin up a `TextWriter` over a `MemoryStream`, serialize to the `TextWriter`, and hand that `MemoryStream` over. Is that about right?

If so, yes that'll probably work, but my concern here is that we *actively don't* want to support arbitrary streams - they can be arbitrary length and "open" (think: a `NetworkStream` that receives data every few seconds). A `MemoryStream` is known to be directly available, well-behaving, etc, so : I think `MemoryStream` could be viable, but it seems to me that what we *actually* want here is support for `ArraySegment<byte>`, or in more recent terms: `Memory<byte>`. We could actually also consider `ReadOnlySequence<byte>`, although that is a bit more exotic.

We can trivially get an `ArraySegment<byte>` from a `MemoryStream` (`TryGetBuffer`), which means that yes, implicit conversion operators from any of `MemoryStream`, `ArraySegment<byte>` or `Memory<byte>` should be simple things to add; we can just replace the existing:

    private readonly byte[] valueBlob;

with:

    private readonly Memory<byte> valueBlob;

(since all of the above are directly and freely convertible to `Memory<byte>`)

So technically I think this is simple and very achievable.

One slight gotcha: I was about to replace `valueBlob` with an `object` to allow me to store `string` and `byte[]`, but I don't want to box - so I guess I'll need to end up with:

        private readonly Memory<byte> valueBlob;
        private readonly long valueInt64;
        private readonly string valueString;

which... I guess is still fine.

Thoughts?

**Comment by mgravell** [comment]
Note: size is `32` bytes (`struct`), which I'm comfortable with; compares to `16` bytes currently, but... meh, this isn't problem. Note: because I was *literally about to make the `string` change*, this is actually a really good time to think about this enhancement.

**Comment by mgravell** [comment]
Implementation footnote: suggest I use **non interned** (this is important) sentinel values in `valueString` and ref-equality; i.e.

```
if((object)valueString == (object)SentinelValueInt64) {
    // trust valueInt64
} else if ((object)valueString == (object)SentinelValueBlob) {
    // trust valueBlob
} else {
    // trust valueString, including "null"
}
```

**Comment by mgravell** [comment]
additional additional; probably want `valueString` to be `valueObject`, that *might happen to be a string*; I'm thinking of `ReadOnlySequence<byte>` - technically a value-type, but quite big; I'm actually content to box that beast

**Comment by mgravell** [comment]
BTW: support for `Memory<byte>` is in the 2.0 pipelines/branch; this means that you can now use the oversized buffer from `MemoryStream` (via `GetBuffer()` / `TryGetBuffer()`) without having to allocate another (exact-sized) array. I **do not** intend to add support for `Stream` in general, for the performance reasons mentioned above.

Does that give you what you need?

**Comment by mgravell** [comment]
@shravan2x when talking about BLOBs, redis has very simple operations - you can assign the entire BLOB, or you can *append* BLOBs. When transferring a single BLOB : at the protocol level, yes: you need to know the length first - the data is length prefixed (it is a "bulk string" in the RESP specification).

Now, in the case of a `Stream`, there are four ways we can think about this:

- pass the `Stream` down to the protocol writer layer send everything at once, writing `Length` bytes (and demanding it start at zero) as a single BLOB - undesirable because a `Stream` has undefined performance characteristics and this could block or break the writer 
- buffer the data at the `Message` layer (happens before the protocol writer), so that we can pass down a completed buffer - undesirable because a `Stream` can have infinite length in addition to undefined performance characteristics, and even if not infinite: can be a significant buffer overhead
- write a loop that is essentially `while((bytes = await stream.ReadAsync(buffer)) > 0) WriteAppendMessage(buffer, bytes);` - undesirable because it is a race condition and is very brittle to corrupted streams; other callers can see incomplete data, or can corrupt the data be appending simultanesously
- demand that the input has known size, is already buffered, and is synchronous and instantaneously available; write those contents directly

Of these, the only one that sounds desirable is the last, and as it happens: `MemoryStream` meets these conditions. As it happens, given your example of `JsonConvert.SerializeObject(...)`, I suspect that you're *already using* `MemoryStream`. Is this the case?

However, in my opinion is it confusing to the consumer for us to take `MemoryStream` directly - because it suggests that we will respect the `Stream` semantics which could change *after* you have added it to the `RedisValue`. So I think that makes support confusion.

First question, then: can you confirm that you are currently using `MemoryStream`?

I can also acknowledge that it is a little awkward and ugly to get from a `MemoryStream` to an `ArraySegment<byte>`, then to a `Memory<byte>` and finally to a `RedisValue`.

I'm willing to help make that easier. Options for consideration:

- we could write an `AsMemory(this MemoryStream stream)` extension method; this seems a little arbitrary *for our library*, though
- we could write a factory method on `RedisValue` - maybe something like `RedisValue CreateFrom(MemoryStream stream)`
- we could write an explicit operator, but I suspect this is a little ... confusing ... in terms of operator semantics

Any thoughts on that?

**Comment by shravan2x** [comment]
Yes, I'm currently using a `MemoryStream`. Besides this, I'm considering options like [Microsoft.IO.RecyclableMemoryStream](https://github.com/Microsoft/Microsoft.IO.RecyclableMemoryStream) to reduce allocations further.

> we could write an AsMemory(this MemoryStream stream) extension method; this seems a little arbitrary for our library, though

> we could write an explicit operator, but I suspect this is a little ... confusing ... in terms of operator semantics

I agree. These are not the cleanest solutions.

> we could write a factory method on `RedisValue` - maybe something like `RedisValue CreateFrom(MemoryStream stream)`

I think this is the best option. It does not impose on the user as much as the other two choices and simply serves as an optional utility method. We should probably also add a warning for users to not write to the `MemoryStream` while we're using it since that might lead to undefined behavior on custom implementations.

P.S. I really appreciate the detailed explanations!

**Comment by mgravell** [comment]
Implemented for 2.0; thanks for the input
